### PR TITLE
Initial commit Hangman

### PR DESCRIPTION
**Pull request by kcisneros**
* Create hangman class
* Create module for display text
* Add list of possible words to github
* Create class to pick a random word

I think roughly this is heading in the right direction. Off the top of my head things I need to do:

- [x] Figure out how to get the "_" on a single line instead of stacked. Currently it looks like this:
```
_
_
_
_
_
```
instead of this: `_ _ _ _ _`
- [x] Figure out how to replace `_` with correct letter, if entered
- [x] Implement functionality where at the start of any turn, add option to save game OR guess a letter
- [x] When program first loads, add an option to load a saved game
- [x] Add a method to ask player if they want to solve the word.. while not explicitly declared in the assignment, it would feel more natural in a game? tbd.. 

The general vibe is there, I think.. this is how it looks like on my terminal:
```
10297-Katherine-Cisneros:lib kcisneros-study$ ruby hangman.rb 
Let's play Hangman!

cambodia
_
_
_
_
_
_
_
_
Enter a letter to solve the secret word: 
z


So far you have guessed: ["z"]

hangman word is: cambodia
Not a match
You have 11 turns left
_
_
_
_
_
_
_
_
Enter a letter to solve the secret word: 
c


So far you have guessed: ["z", "c"]

hangman word is: cambodia
guess word is: c
You have 10 turns left
```

For testing purposes I'm displaying the hangman word, but won't actually do it once it is working.

### CODE REVIEW

**Comment by kcisneros** [comment]
Not a fan of the game play. I hate asking all these questions, but that's how it is in the game IRL I suppose. It just feels too clunky as a player, but things can't be seamless on the screen. If you want to play the game, give it a go here: https://replit.com/@kcisnerosreplit/RelievedHumbleVaporware

File: ruby_oop/hangman/lib/random_word.rb [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+    @rand_word = random_word
+  end
+
+  def random_word
+    loop do
+      @rand_word = File.readlines('../assets/english_words.txt').sample
+      break if @rand_word.length.between?(6, 12)
+    end
+    @rand_word
+  end
```
Comment (zacharydanger) [inline ruby_oop/hangman/lib/random_word.rb:18]: Both of these methods mutate the `@rand_word` ivar. It makes a lot of sense to set it in `#initalize`, but in `#random_word` it feels more like a side-effect. Try having `#random_word` return the word without touching any ivar.

File: ruby_oop/hangman/lib/random_word.rb [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+
+  def initialize
+    @rand_word = random_word
+  end
+
+  def random_word
+    loop do
+      @rand_word = File.readlines('../assets/english_words.txt').sample
+      break if @rand_word.length.between?(6, 12)
+    end
```
Comment (zacharydanger) [inline ruby_oop/hangman/lib/random_word.rb:16]: `File.readlines` is a relatively expensive operation. And in this case we aren't expecting it to return different results each iteration, so we can pull it out of the `loop`.

```ruby
word_list = File.readlines('...')
loop do
  word = word_list.sample
  # ...
end
```

**Review by Kaktug — COMMENTED** [review]
I left a few minor comments for your consideration.  I know you mentioned that you dislike the number of questions being asked...what if you did something like the outline below?

In `prompt_for_letter`, you are storing whatever the player types into the `@prompt_guess_letter` var.  What if you checked that var for key words/phrases, such as 'save game' before running `display_guessed_letters` and the other logic under the `each_turn` method?  Something like:

```
# hangman.rb

def each_turn
  prompt_for_letter # this updates @prompt_guess_letter
  save_the_game_option
  [...]
end
```

```
# manage_game_state.rb

def save_the_game_option
  if @turn_number <= 11 && !win?
  case @prompt_guess_letter
  when 'save game'
    puts DisplayableText::GAME_SAVED_CONFIRMATION
    save_game_file
   exit(0)
  end
end
```

Then at each turn the prompt could read something like:

```
Let's play Hangman!
-------- Word to guess is 8 letters long!
Enter a letter to solve the secret word, or type 'save game' to save and quit:
```

That way you wouldn't have to ask the player if they want to save & quit at each turn.  They can tell the game to do it themselves, whenever they want.

File: ruby_oop/hangman/lib/manage_game_state.rb [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+  def save_game_file
+    serialized_object = File.new(FILE_NAME, 'w')
+    serialized_object.puts YAML.dump(self)
+    serialized_object.close
+  end
+
+  def save_the_game_option
+    if @turn_number <= (Hangman::MAX_TURNS - 1) && !win?
+      case @prompt_for_letter
+      when 'save game'
```
Comment (chmurph2) [inline ruby_oop/hangman/lib/manage_game_state.rb:40]: It's a little odd to have a `case` with just one `when`. Any reason not to use `if @prompt_for_letter == 'save game'`?

Also, it's good practice use a constant for load-bearing strings like `'save game'`.

**Review by chmurph2 — APPROVED** [review]
You already received a lot of good feedback at this point. I had a couple minor suggestions, but LGTM.

File: ruby_oop/hangman/lib/manage_game_state.rb (hunk 3 above)
Comment (kcisneros) [reply ruby_oop/hangman/lib/manage_game_state.rb:40]: >It's a little odd to have a case with just one when. Any reason not to use if @prompt_for_letter == 'save game'?
I _think_ rubocop complained when I had it as an `if` and proposed a `case`, but maybe I am mixing up rubocop suggestions with personal style choices and it's all a blur in my head. 

I'll update the string to a constant though.

**Pull request merged**